src/components/rangeslider/draw.js

```diff
@@ -0,0 +1,424 @@
+};
+
+function makeRangeSliderData(fullLayout) {
+    if(!fullLayout.xaxis) return [];
+    if(!fullLayout.xaxis[constants.name]) return [];
+    if(!fullLayout.xaxis[constants.name].visible) return [];
+    if(fullLayout._has('gl2d')) return [];
+
+    return [fullLayout.xaxis];
+}
+
+function setupDragElement(rangeSlider, gd, axisOpts, opts) {
+    var slideBox = rangeSlider.select('rect.' + constants.slideBoxClassName).node(),
+        grabAreaMin = rangeSlider.select('rect.' + constants.grabAreaMinClassName).node(),
+        grabAreaMax = rangeSlider.select('rect.' + constants.grabAreaMaxClassName).node();
+
+    rangeSlider.on('mousedown', function() {
+        var event = d3.event,
+            target = event.target,
+            startX = event.clientX,
+            offsetX = startX - rangeSlider.node().getBoundingClientRect().left,
+            minVal = opts.d2p(axisOpts.range[0]),
+            maxVal = opts.d2p(axisOpts.range[1]);
+
+        var dragCover = dragElement.coverSlip();
+
+        dragCover.addEventListener('mousemove', mouseMove);
+        dragCover.addEventListener('mouseup', mouseUp);
+
+        function mouseMove(e) {
+            var delta = +e.clientX - startX;
+            var pixelMin, pixelMax, cursor;
+
+            switch(target) {
+                case slideBox:
+                    cursor = 'ew-resize';
+                    pixelMin = minVal + delta;
+                    pixelMax = maxVal + delta;
+                    break;
+
+                case grabAreaMin:
+                    cursor = 'col-resize';
+                    pixelMin = minVal + delta;
+                    pixelMax = maxVal;
+                    break;
+
+                case grabAreaMax:
+                    cursor = 'col-resize';
+                    pixelMin = minVal;
+                    pixelMax = maxVal + delta;
+                    break;
+
+                default:
+                    cursor = 'ew-resize';
+                    pixelMin = offsetX;
+                    pixelMax = offsetX + delta;
+                    break;
+            }
+
+            if(pixelMax < pixelMin) {
+                var tmp = pixelMax;
+                pixelMax = pixelMin;
+                pixelMin = tmp;
+            }
+
+            opts._pixelMin = pixelMin;
+            opts._pixelMax = pixelMax;
+
+            setCursor(d3.select(dragCover), cursor);
+            setDataRange(rangeSlider, gd, axisOpts, opts);
+        }
+
+        function mouseUp() {
+            dragCover.removeEventListener('mousemove', mouseMove);
+            dragCover.removeEventListener('mouseup', mouseUp);
+            Lib.removeElement(dragCover);
+        }
+    });
+}
+
+function setDataRange(rangeSlider, gd, axisOpts, opts) {
+
+    function clamp(v) {
+        return Lib.constrain(v, opts.range[0], opts.range[1]);
+    }
+
+    var dataMin = clamp(opts.p2d(opts._pixelMin)),
+        dataMax = clamp(opts.p2d(opts._pixelMax));
+
+    window.requestAnimationFrame(function() {
```

Comment:
I'm sure it's valid, but I'm curious what effect `requestAnimationFrame` has here. I'm not sure I've ever really used it without calling is pseudo-recursively. Is the goal to get it bumped down in priority below user interactions so that relayout doesn't block the UI?